File: implementations/go/provekit-ir-symbolic/verifier/instantiate.go
package verifier

import "fmt"

// InstantiateStage substitutes the call-site argument term into the
// outermost forall of the resolved precondition. The result is the
// per-callsite IR-formula obligation the solver discharges.
//
//	forall n: Int. n > 0    instantiated at Const 0
//	→                       0 > 0
//
// v1.1.0 IR shape: quantifier is FLAT; {kind, name, sort, body} with
// no Lambda wrapper. The bound variable's identifier is the top-level
// `name` field (was `predicate.varName` pre-v1.1.0).
type InstantiateStage struct{}

// Run drops the outermost forall and substitutes its bound variable
// with `arg` throughout the body. Returns the substituted formula or
// an error if the input isn't a forall.
func (s *InstantiateStage) Run(formula interface{}, arg interface{}) (interface{}, error) {
	f, ok := formula.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("instantiate: formula is not a map")
	}
	if f["kind"] != "forall" {
		return nil, fmt.Errorf("instantiate: outermost is %v, expected forall", f["kind"])
	}
	varName, _ := f["name"].(string)
	if varName == "" {
		return nil, fmt.Errorf("instantiate: forall has empty bound-variable name")
	}
	body, ok := f["body"].(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("instantiate: forall body missing")
	}
	return substituteFormula(body, varName, arg), nil
}

// substituteFormula walks a formula tree (v1.1.0 shape) and replaces
// var references matching `name` with `replacement` (a JSON-shape value
// of an IrTerm).
func substituteFormula(f map[string]interface{}, name string, replacement interface{}) map[string]interface{} {
	out := cloneMap(f)
	switch f["kind"] {
	case "atomic":
		if args, ok := f["args"].([]interface{}); ok {
			newArgs := make([]interface{}, len(args))
			for i, a := range args {
				if am, ok := a.(map[string]interface{}); ok {
					newArgs[i] = substituteTerm(am, name, replacement)
				} else {
					newArgs[i] = a
				}
			}
			out["args"] = newArgs
		}
	case "and", "or", "not", "implies":
		if operands, ok := f["operands"].([]interface{}); ok {
			newOps := make([]interface{}, len(operands))
			for i, op := range operands {
				if om, ok := op.(map[string]interface{}); ok {
					newOps[i] = substituteFormula(om, name, replacement)
				} else {
					newOps[i] = op
				}
			}
			out["operands"] = newOps
		}
	case "forall", "exists":
		// Shadowing: don't substitute past a binder that re-introduces `name`.
		if f["name"] == name {
			return out
		}
		if body, ok := f["body"].(map[string]interface{}); ok {
			out["body"] = substituteFormula(body, name, replacement)
		}
	case "choice":
		// Shadowing: don't substitute past a binder that re-introduces `name`.
		if f["varName"] == name {
			return out
		}
		if body, ok := f["body"].(map[string]interface{}); ok {
			out["body"] = substituteFormula(body, name, replacement)
		}
	}
	return out
}

func substituteTerm(t map[string]interface{}, name string, replacement interface{}) interface{} {
	if t["kind"] == "var" && t["name"] == name {
		return replacement
	}
	if t["kind"] == "ctor" {
		out := cloneMap(t)
		if args, ok := t["args"].([]interface{}); ok {
			newArgs := make([]interface{}, len(args))
			for i, a := range args {
				if am, ok := a.(map[string]interface{}); ok {
					newArgs[i] = substituteTerm(am, name, replacement)
				} else {
					newArgs[i] = a
				}
			}
			out["args"] = newArgs
		}
		return out
	}
	if t["kind"] == "lambda" {
		// Shadowing: don't substitute past a binder that re-introduces `name`.
		if t["paramName"] == name {
			return t
		}
		out := cloneMap(t)
		if body, ok := t["body"].(map[string]interface{}); ok {
			out["body"] = substituteTerm(body, name, replacement)
		}
		return out
	}
	if t["kind"] == "let" {
		out := cloneMap(t)
		shadowed := false
		if bindings, ok := t["bindings"].([]interface{}); ok {
			newBindings := make([]interface{}, len(bindings))
			for i, b := range bindings {
				if bm, ok := b.(map[string]interface{}); ok {
					if !shadowed {
						newBoundTerm := bm["boundTerm"]
						if btm, ok := bm["boundTerm"].(map[string]interface{}); ok {
							newBoundTerm = substituteTerm(btm, name, replacement)
						}
						newBindings[i] = map[string]interface{}{
							"name":      bm["name"],
							"boundTerm": newBoundTerm,
						}
						if bm["name"] == name {
							shadowed = true
						}
					} else {
						newBindings[i] = b
					}
				} else {
					newBindings[i] = b
				}
			}
			out["bindings"] = newBindings
		}
		if !shadowed {
			if body, ok := t["body"].(map[string]interface{}); ok {
				out["body"] = substituteTerm(body, name, replacement)
			}
		}
		return out
	}
	return t
}

func cloneMap(m map[string]interface{}) map[string]interface{} {
	out := make(map[string]interface{}, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

File: implementations/rust/provekit-verifier/src/instantiate.rs
// SPDX-License-Identifier: Apache-2.0
//
// Stage 4: instantiate. Substitute the call's arg term for the
// resolved forall's bound variable. Flat quantifier shape: the
// resolved formula is expected to be `{kind:"forall", name, sort, body}`;
// we substitute `arg_term` for `name` in `body`.
//
// Mirrors .../verifier/instantiate.cpp.

use serde_json::{json, Value as Json};

use crate::types::{Obligation, ResolvedProperty};

pub fn run(resolved: &ResolvedProperty, arg_term: &Option<Json>) -> Result<Obligation, String> {
    let arg = arg_term
        .as_ref()
        .ok_or("no argument term to substitute")?;
    let f = resolved
        .ir_formula
        .as_ref()
        .ok_or("resolved property has no ir_formula (no pre slot)")?;
    if f.get("kind").and_then(|v| v.as_str()) != Some("forall") {
        return Err("precondition formula is not a forall".into());
    }
    let var_name = f
        .get("name")
        .and_then(|v| v.as_str())
        .ok_or("forall has empty bound-variable name")?;
    let sort = f
        .get("sort")
        .ok_or("forall has no sort")?
        .clone();
    let body = f
        .get("body")
        .ok_or("forall has no body")?;
    let substituted_body = substitute_formula(body, var_name, arg);
    let forall_with_sort = json!({
        "kind": "forall",
        "name": var_name,
        "sort": sort,
        "body": substituted_body
    });
    Ok(Obligation {
        property_cid: resolved.cid.clone(),
        ir_kit_version: resolved.ir_kit_version.clone(),
        ir_formula: forall_with_sort,
    })
}

/// Public adapter so other stages (notably the handshake's
/// implication-form obligation builder) can reuse the same
/// alpha-renaming helper.
pub fn substitute_formula_pub(f: &Json, name: &str, replacement: &Json) -> Json {
    substitute_formula(f, name, replacement)
}

fn substitute_formula(f: &Json, name: &str, replacement: &Json) -> Json {
    let mut out = f.clone();
    let kind = f.get("kind").and_then(|v| v.as_str()).unwrap_or_default();
    if let Json::Object(map) = &mut out {
        match kind {
            "atomic" => {
                if let Some(Json::Array(args)) = map.get("args").cloned() {
                    let new_args: Vec<Json> = args
                        .iter()
                        .map(|a| substitute_term(a, name, replacement))
                        .collect();
                    map.insert("args".into(), Json::Array(new_args));
                }
            }
            "and" | "or" | "not" | "implies" => {
                if let Some(Json::Array(ops)) = map.get("operands").cloned() {
                    let new_ops: Vec<Json> = ops
                        .iter()
                        .map(|op| substitute_formula(op, name, replacement))
                        .collect();
                    map.insert("operands".into(), Json::Array(new_ops));
                }
            }
            "forall" | "exists" => {
                let bound = map.get("name").and_then(|v| v.as_str()).unwrap_or_default();
                if bound == name {
                    // Shadowed; do not descend.
                    return out;
                }
                if let Some(body) = map.get("body").cloned() {
                    map.insert("body".into(), substitute_formula(&body, name, replacement));
                }
            }
            _ => {}
        }
    }
    out
}

fn substitute_term(t: &Json, name: &str, replacement: &Json) -> Json {
    if !t.is_object() {
        return t.clone();
    }
    let kind = t.get("kind").and_then(|v| v.as_str()).unwrap_or_default();
    if kind == "var" && t.get("name").and_then(|v| v.as_str()) == Some(name) {
        return replacement.clone();
    }
    if kind == "ctor" {
        let mut out = t.clone();
        if let Json::Object(map) = &mut out {
            if let Some(Json::Array(args)) = map.get("args").cloned() {
                let new_args: Vec<Json> = args
                    .iter()
                    .map(|a| substitute_term(a, name, replacement))
                    .collect();
                map.insert("args".into(), Json::Array(new_args));
            }
        }
        return out;
    }
    t.clone()
}
